### PR TITLE
UI, Input Date

### PR DESCRIPTION
**Pull request by nhaagen**
Date uses bootstrap datetimepicker to select a date or time (or both).

bildschirmfoto_2018-08-13_16-33-45

Rules for the glyphs:
https://github.com/ILIAS-eLearning/ILIAS/pull/1142/files#diff-636dc2f66a0722b4e790c5654824ddc1

Rules for the input:
https://github.com/ILIAS-eLearning/ILIAS/pull/1142/files#diff-2d8571e3a451f51b51f9868081f6f7e8

For time-ranges, use the Duration input (which  is a group of dates):
bildschirmfoto_2018-08-14_14-53-10

### CODE REVIEW

File: src/Transformation/Transformations/Date.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+class Date implements Transformation {
+
+	/**
+	 * @inheritdoc
+	 */
+	public function transform($from) {
+		if($from) {
+			return new \DateTime($from);
+		}
+		return '';
```
Comment (klees) [inline src/Transformation/Transformations/Date.php]: This is bad. The transformation promises to return a \DateTime, this ain't one. Please just throw an `\InvalidArgumentException` if the input to `transform` is not as expected.

File: src/UI/Component/Input/Field/Duration.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+namespace ILIAS\UI\Component\Input\Field;
+
+/**
+ * This describes the duration input.
+ */
+interface Duration extends Group { //extend group? or extend date?
+
+	/**
+	 * Get an input like this using the given format.
+	 */
```
Comment (klees) [inline src/UI/Component/Input/Field/Duration.php:20]: What is a valid format here? I certainly can't just pass _any_ string, right? Is it anything related to phps `\DateTime`? Just say so...

File: src/UI/Component/Input/Field/Duration.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	public function withFormat(string $format) : Duration;
+
+	/**
+	 * Return the input's date-format
+	 */
+	public function getFormat() : string;
+
+	/**
+	 * Limit accepted values to dates past given $date.
```
Comment (klees) [inline src/UI/Component/Input/Field/Duration.php]: _Past given date_ means given date is excluded?

File: src/UI/Component/Input/Field/Duration.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	 */
+	public function withMinDate(\DateTime $date) : Duration;
+
+	/**
+	 * Return the lowest date the input accepts.
```
Comment (klees) [inline src/UI/Component/Input/Field/Duration.php]: This to me implies _including_ the given date.

File: src/UI/Component/Input/Field/Duration.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	 * @return  \DateTime | null
+	 */
+	public function getMinDate();
+
+	/**
+	 * Limit accepted values to dates before given $date.
```
Comment (klees) [inline src/UI/Component/Input/Field/Duration.php]: I think we have the same including/excluding problem here.

File: src/UI/Implementation/Component/Input/Field/Date.php [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+		$trafo = $transformation_factory->toDate();
+		$this->setAdditionalTransformation($trafo);
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function withFormat(string $format) : C\Input\Field\Date {
+		$this->checkStringArg('format', $format); //2do: check on date-format
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/Date.php]: Do it!

File: src/UI/Implementation/Component/Input/Field/Date.php [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	public function getTimeGlyph() : bool {
+		return $this->use_time_glyph;
+	}
+
+
+	/**
+	 * @inheritdoc
+	 */
+	protected function isClientSideValueOk($value) {
+		return is_string($value);
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/Date.php]: The client side delivers values as strings according to the format, right? This should be reflected here.

File: src/UI/Implementation/Component/Input/Field/Duration.php [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+	}
+
+
+	protected function addTransformation(TransformationFactory $transformation_factory) {
+		$duration = $transformation_factory->custom(function($v) {
+			list($from, $until) = $v;
+			if($from && $until) {
+				return ['start'=>$from, 'end'=>$until, 'interval'=>$from->diff($until)];
+			}
+			return '';
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/Duration.php]: This is odd. So the field either contains a `\DateDuration` or an empty string? Please throw here if input is not what you expected.

File: src/UI/Implementation/Component/Input/Field/Duration.php [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+	protected function addValidation(ValidationFactory $validation_factory) {
+		$from_before_until = $validation_factory->custom(
+			function($v) {
+				if($v === '') {
+					return true;
+				}
+				return $v['start'] < $v['end'];
+			},
+			"start must be earlier than end"
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/Duration.php]: Please add me a marker TODO: i18n here to i18n this.

File: src/UI/Implementation/Component/Input/Field/Duration.php [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+		$this->setAdditionalTransformation($duration);
+	}
+
+	protected function addValidation(ValidationFactory $validation_factory) {
+		$from_before_until = $validation_factory->custom(
+			function($v) {
+				if($v === '') {
+					return true;
+				}
+				return $v['start'] < $v['end'];
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/Duration.php]: These are `\DateTime`s here, right? Does the `<`-Operator treat them properly?

**Review by klees — CHANGES_REQUESTED** [review]
LGTM in general, THX!!!

Please have a look at your `Transformation`s, they ought to throw, not silently forward empty strings if input is not what is expected.

File: src/UI/Component/Input/Field/Duration.php (hunk 2 above)
Comment (Amstutz) [reply src/UI/Component/Input/Field/Duration.php:20]: maybe even add some link, like: http://php.net/manual/de/datetime.format.php or http://php.net/manual/de/class.datetime.php

File: src/UI/Component/Input/Field/Date.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+/* Copyright (c) 2018 Nils Haagen <nils.haagen@concepts-and-training.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Component\Input\Field;
+
+/**
+ * This describes the date-field.
+ */
+interface Date extends Input {
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Date.php]: Is it not DateTime? To me Date only excludes time (which is part of the conrol).

File: src/UI/Component/Input/Field/Date.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+	/**
+	 * Get an input like this using the given format.
+	 */
+	public function withFormat(string $format) : Date;
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Date.php]: This method is of crucial importance, it defines the kind of data that is accepted here (time, dates, both). This should be made more explicit here. Add some more explanation. Maybe even some little dummy example.

File: src/UI/Component/Input/Field/Factory.php [hunk 13]
```diff
@@ -348,5 +348,64 @@ public function password($label, $byline = null);
 	 * @return \ILIAS\UI\Component\Input\Field\Select
 	 */
 	public function select($label, array $options, $byline = null);
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     A Date Input is used to enter dates.
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: and/or time.

File: src/UI/Component/Input/Field/Factory.php [hunk 14]
```diff
@@ -348,5 +348,64 @@ public function password($label, $byline = null);
...
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     A Duration Input is used to enter a time span instead of a single point in time.
+	 *   composition: >
+	 *     A Duration Input is composed as a group of two Date Inputs.
+	 *   effect: >
+	 *     Dependant on the format, the inputs will accept dates, times or datetimes.
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php:528]: There is something missing like:

The start point must take place before the endpoint. Invalid input will be corrected automatically.

**Review by Amstutz — CHANGES_REQUESTED** [review]
The examples look great. Thx a lot for this @nhaagen. There is one major point: I believe the name "date" to be misleading, should be "DateTime" since it is possible to enter both (also, closer to the PHP naming of the according classes for the data types). The rest is minor stuff. We can discuss all of this on monday.

One thing I believe to be also slighty interface related is the behaviour off the Duration Input if end is set smaller than start. I would wish this to be impossible. Currently it is (in the example). If we can not prevent this for some reason, I think we should a least correct autamitcally (e.g. by setting end to start). We should discuss this point also in the JF. Note in the examples of the library, this is impossible to enter wrongly: https://eonasdan.github.io/bootstrap-datetimepicker/#linked-pickers

**Comment by klees** [comment]
Removed JF label to clean up issues first.

**Comment by Amstutz** [comment]
Is this in a deadlock state? Anything needed from my part?

**Comment by klees** [comment]
No deadlock, just no time to clean up issues till now.

**Comment by nhaagen** [comment]
Hi @Amstutz ,
it certainly took a while, but now - from my perspective - this is good to go.
Would you have a look?

**Comment by Amstutz** [comment]
Hi @nhaagen 

Thanks a lot for the revision. As far as I am concerned, all my remarks on the interface are taken care of. To me, this is JF ready, I'll attached the according label.

Note, that I will give a final review on the implementation as soon as the interface is accepted by the JF.

Thx a lot.
@Amstutz

**Comment by klees** [comment]
Hi everybody, I'm still no happy with the current state of this, so I removed the JF label. We will get back to this soon try to present this on JF 2019-02-04.

**Comment by Amstutz** [comment]
Any Update? Note that this is the oldest PR by now ;-)

**Comment by Amstutz** [comment]
There has been no update for almost three month on this, to keep the list of PR's clean, Ill close this down for the moment. You may reopen this or make a new PR as soon as there is something new to discuss.

**Pull request closed without merging**